src/themis/secure_message.c

```diff
@@ -77,6 +77,74 @@ themis_status_t themis_gen_ec_key_pair(uint8_t* private_key,
   return themis_gen_key_pair(SOTER_SIGN_ecdsa_none_pkcs8, private_key, private_key_length, public_key, public_key_length);
 }
 
+themis_status_t themis_secure_message_encrypt(const uint8_t* private_key,
+                                              const size_t private_key_length,
+                                              const uint8_t* public_key,
+                                              const size_t public_key_length,
+                                              const uint8_t* message,
+                                              const size_t message_length,
+                                              uint8_t* encrypted_message,
+                                              size_t* encrypted_message_length)
+{
+  THEMIS_CHECK_PARAM(private_key!=NULL);
+  THEMIS_CHECK_PARAM(private_key_length!=0);
+  THEMIS_CHECK_PARAM(public_key!=NULL);
+  THEMIS_CHECK_PARAM(public_key_length!=0);
+  THEMIS_CHECK_PARAM(message!=NULL);
+  THEMIS_CHECK_PARAM(message_length!=0);
+  THEMIS_CHECK_PARAM(encrypted_message_length!=NULL);
+  return themis_secure_message_wrap(private_key, private_key_length, public_key, public_key_length, message, message_length, encrypted_message, encrypted_message_length);
+}
+
+themis_status_t themis_secure_message_decrypt(const uint8_t* private_key,
+                                              const size_t private_key_length,
+                                              const uint8_t* public_key,
+                                              const size_t public_key_length,
+                                              const uint8_t* encrypted_message,
+                                              const size_t encrypted_message_length,
+                                              uint8_t* message,
+                                              size_t* message_length)
+{
+  THEMIS_CHECK_PARAM(private_key!=NULL);
+  THEMIS_CHECK_PARAM(private_key_length!=0);
+  THEMIS_CHECK_PARAM(public_key!=NULL);
+  THEMIS_CHECK_PARAM(public_key_length!=0);
+  THEMIS_CHECK_PARAM(encrypted_message!=NULL);
+  THEMIS_CHECK_PARAM(encrypted_message_length!=0);
+  THEMIS_CHECK_PARAM(message_length!=NULL);
+  return themis_secure_message_unwrap(private_key, private_key_length, public_key, public_key_length, encrypted_message, encrypted_message_length, message, message_length);
+}
+
+themis_status_t themis_secure_message_sign(const uint8_t* private_key,
```

Comment:
maybe it's time to verify the type of key to avoid misusage incorrect type of key (public/private)? : ) 
we can do simple check on prefix of key value for known prefixes of private keys:
* rsa private key: https://github.com/cossacklabs/themis/blob/master/src/soter/soter_rsa_key.h#L25
* ec private key: https://github.com/cossacklabs/themis/blob/master/src/soter/soter_ec_key.h#L26
```
if (memcmp(private_key, RSA_PRIV_KEY_PREF, sizeof(RSA_PRIV_KEY_PREF) != 0 && 
    memcmp(private_key, RSA_PRIV_KEY_PREF, sizeof(EC_PRIV_KEY_PREF) != 0) { 
return THEMIS_INVALID_PARAMETER;
}
```

Reply:
but better will be to extend api and add some functions that will validate keys: `validate_private_key`/`validate_public_key` and for now, it will check the only prefix. In the future we can extend these validations with verifying full key's structure

Reply:
Come to think of it... I've added the following utilities for the Rust wrapper to query the key kind and validate its content:

https://github.com/cossacklabs/themis/blob/29ebf63e968f036f8d2fdd0a9d458a4ae7f84728/src/wrappers/themis/rust/libthemis-sys/src/wrapper.h#L28-L41

Can they be of some use? What do you think of moving these functions to the core library (and using them for checks internally as well)? I believe this API may be useful for applications to verify key content independently.

As for Secure Messages, I guess we can even go a bit further and verify that both private and public keys are of ECDSA or RSA flavor. Ideally, we'd also verify that both use the same public parameters, but that's harder to do.

Reply:
good checks, I think it's good to move into the core themis library

Reply:
agree, good checks, make sense at core lvl

Reply:
@Lagovas @vixentael I've moved the key validation routines to Themis core, added some tests for them, added validation to new Secure Message API, and added some early checks to ThemisPP and JsThemis which are most affected by poor error reporting. Please give it a look.